Add item requests

Believerin opened the pull request.

Believerin committed:
Initial commit

Believerin committed:
Initial commit

Believerin committed:
Initial commit

Believerin committed:
perf: "проверка на дублирование почты изменена и перенесена в UserDbStorage"
fix: "в метод getItem добавлена обработка null"
feat: "в dto добавлены аннотации валидации"
feat: "аннотация @Component заменёна на @Repository"

Believerin committed:
feat: "добавлена функция бронирования и комментирования"

Believerin committed:
feat: "добавлена функция бронирования и комментирования"

Believerin committed:
perf: "Убраны лишние обращения к БД"
refactor: "Добавлены аннотации @Transactional"
refactor: "Методы вынесены из маппера в сервис"

Believerin committed:
perf: "Убраны лишние обращения к БД"
refactor: "Добавлены аннотации @Transactional"
refactor: "Методы вынесены из маппера в сервис"

Believerin committed:
perf: "Убраны лишние обращения к БД"
refactor: "Убрано возвращение лишних выходных данных"
refactor: "Удалено дублирование кода извлечением в метод"

Believerin committed:
feature: "Добавлены классы ResponseDto И RequestDto"
feature: "Добавлены классы для полученияи возвращения Item"
feature: "Добавлены классы для полученияи возвращения Booking"

Believerin committed:
Raw code

Believerin committed:
"feat: Добавлена обработка сущности Request"
"test: Создано юнит тесты сервисных слоёв"
"test: Создано по одному интеграционному тесту сервисных слоёв"
"test: Созданы тесты вэб-слоёв контроллеров"
"test: Созданы тесты DataTest"
"test: Созданы тесты JsonTest для возвращаемых Dto"

Believerin committed:
"feat: Добавлена обработка сущности Request"
"test: Создано юнит тесты сервисных слоёв"
"test: Создано по одному интеграционному тесту сервисных слоёв"
"test: Созданы тесты вэб-слоёв контроллеров"
"test: Созданы тесты DataTest"
"test: Созданы тесты JsonTest для возвращаемых Dto"

Believerin committed:
"feat: Добавлена обработка сущности Request"
"test: Создано юнит тесты сервисных слоёв"
"test: Создано по одному интеграционному тесту сервисных слоёв"
"test: Созданы тесты вэб-слоёв контроллеров"
"test: Созданы тесты DataTest"
"test: Созданы тесты JsonTest для возвращаемых Dto"

Believerin committed:
"feat: Добавлена обработка сущности Request"
"test: Создано юнит тесты сервисных слоёв"
"test: Создано по одному интеграционному тесту сервисных слоёв"
"test: Созданы тесты вэб-слоёв контроллеров"
"test: Созданы тесты DataTest"
"test: Созданы тесты JsonTest для возвращаемых Dto"

Alexander-Bolotov commented on src/main/java/ru/practicum/shareit/request/model/Request.java:
```diff
@@ -0,0 +1,29 @@
+    @JoinColumn(name = "user_id")
+    private User requester;
+    @Column(name = "creation_date")
+    private LocalDateTime created;
```
Можно использовать     @CreationTimestamp тогда если даты не будет, то она автоматически создастся при создании записи

Alexander-Bolotov commented on src/main/java/ru/practicum/shareit/request/dto/ResponseDtoCreated.java:
```diff
@@ -0,0 +1,24 @@
+@RequiredArgsConstructor
+@AllArgsConstructor
+@Builder
+public class ResponseDtoCreated {
```
а есть смысл в двух подобных дто, я так понимаю отличие от ResponceDto только в одно поле. Нельзя обойтись одной ДТО?

Believerin replied:
Ревьюер на замене указывал на создание DTO с целью возвращения только требуемых данных, даже с отличием лишь в одно поле.

Alexander-Bolotov replied:
Тоже есть своя логика) ок

Alexander-Bolotov commented on src/main/java/ru/practicum/shareit/request/dto/ResponseDto.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Builder
+    @Data
+    public static class ItemOffer {
```
лучше, думаю, вынести в отдельное дто. Да и наверное хорошо бы хранить эти ItemOfer в БД тоже

Alexander-Bolotov commented on src/main/java/ru/practicum/shareit/request/RequestRepository.java:
```diff
@@ -0,0 +1,13 @@
+package ru.practicum.shareit.request;
```
Репозитории лучше тоже помещать в отдельный пекедж

Alexander-Bolotov commented on src/main/java/ru/practicum/shareit/request/service/RequestServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+    private final UserRepository userRepository;
+    private final UserService userService;
+
+    private Function<Integer, List<Item>> findItemsByRequestId;
```
Сервис не должен хранить состояние (да в принципе, как по хорошему и любой другой класс)
Не нужно хранить содержание, каких-то запросов. Делай их там, где они нужны будут

Alexander-Bolotov commented on src/main/java/ru/practicum/shareit/user/UserMapper.java:
```diff
@@ -0,0 +1,24 @@
+import ru.practicum.shareit.user.model.User;
+
+@UtilityClass
+public class UserMapper {
```
Можно воспользоваться MapStruct для генерации мапперов

Believerin committed:
"feat: Добавлена @CreationTimestamp в Request"
"feat: Создана таблица для ItemOffer"
"feat: Для репозиториев созданы отдельные пакеты"
"refactor: Переменные Функциональных интерфейсов перенесыны в вызываемые методы"
"refactor: Dto для User генерируется с помощью MapStruct"
